exporter/exporter.go

```diff
@@ -153,8 +156,39 @@ func (e *Exporter) scrape(groupScrapes chan<- GroupScrapeResult) {
 	var errorCount uint64 = 0
 
 	asgSvc := autoscaling.New(e.session, aws.NewConfig())
+	// get all AutoScalingGroupNames
+	groups, filterErr := asgSvc.DescribeAutoScalingGroups(&autoscaling.DescribeAutoScalingGroupsInput{})
```

Comment:
The way this is implemented means that you are trying to request all groups, then filter the returned list based on a provided regex, before then re-requesting the data for the selected groups (which you would already have from this call here).

Also note that I wrote "trying to": you are not using the `.DescribeAutoScalingGroupsPages` helper, which means you will only get one page of results, which could leave some ASGs to be never found.

Overall you can reduce the amount of work and API requests done and simplify the code itself by just applying the filter within the callback-function in the call to `.DescribeAutoScalingGroupsPages`: instead of unconditionally invoking the dynamically created `go func` with the `asg`, just apply the filter right then and just `continue` if the ASG at hand should not be checked.

In essence:

```go
err := asgSvc.DescribeAutoScalingGroupsPages(&autoscaling.DescribeAutoScalingGroupsInput{
	AutoScalingGroupNames: aws.StringSlice(groupsList),
}, func(result *autoscaling.DescribeAutoScalingGroupsOutput, lastPage bool) bool {
	// ...
	for _, asg := range result.AutoScalingGroups {
		if e.groupsFilter != "" && asg.AutoScalingGroupName != nil {
			matched, err := regexp.MatchString(e.groupsFilter, *asg.AutoScalingGroupName)
			if err != nil { /* ... */ }
			if !matched {
				continue
			}
		}
		// ...
	}
	// ...
})
```

Reply:
I hope, the current state is what you had in mind.